File: src/lae/util/xdg.py
"""XDG path helpers."""

from __future__ import annotations

import os
from pathlib import Path


def expand(path: str | Path) -> Path:
    return Path(os.path.expanduser(str(path))).resolve()


def config_home() -> Path:
    return expand(os.environ.get("XDG_CONFIG_HOME", "~/.config"))


def data_home() -> Path:
    return expand(os.environ.get("XDG_DATA_HOME", "~/.local/share"))


def runtime_dir() -> Path:
    base = os.environ.get("XDG_RUNTIME_DIR")
    if not base:
        raise RuntimeError("XDG_RUNTIME_DIR is not set")
    return Path(base)


def lae_config_dir() -> Path:
    return config_home() / "lae"


def lae_config_path() -> Path:
    return lae_config_dir() / "config.toml"


def lae_data_dir() -> Path:
    return data_home() / "lae"


def lae_state_db() -> Path:
    return lae_data_dir() / "state.db"


def lae_runtime_dir() -> Path:
    return runtime_dir() / "lae"


def resolve_daemon_socket(configured: str) -> Path:
    """Resolve `[daemon].socket` from config to an absolute path."""
    value = configured.strip()
    if value.startswith("~") or value.startswith("/"):
        return expand(value)
    if value in {"daemon.sock", "lae/daemon.sock"}:
        return lae_data_dir() / "daemon.sock"
    return lae_data_dir() / value


def lae_daemon_socket() -> Path:
    from lae.core.config import load_config

    return resolve_daemon_socket(load_config().daemon_socket)


def lae_context_file() -> Path:
    return lae_runtime_dir() / "context"


def lae_waybar_file() -> Path:
    return lae_runtime_dir() / "waybar.json"


def share_dir() -> Path:
    """Shipped static templates (repo share/ or installed package)."""
    pkg = Path(__file__).resolve().parent.parent
    candidate = pkg.parent.parent / "share"
    if candidate.is_dir():
        return candidate
    import importlib.resources as ir

    try:
        ref = ir.files("lae").joinpath("../../share")
        resolved = Path(str(ref))
        if resolved.is_dir():
            return resolved
    except (TypeError, FileNotFoundError):
        pass
    return lae_data_dir()

File: crates/lae-core/src/xdg.rs
use std::env;
use std::path::{Path, PathBuf};

use crate::error::{LaeError, Result};

pub fn expand(path: impl AsRef<Path>) -> PathBuf {
    let s = path.as_ref().to_string_lossy();
    if s.starts_with('~') {
        if let Ok(home) = env::var("HOME") {
            return PathBuf::from(s.replacen('~', &home, 1));
        }
    }
    PathBuf::from(s.into_owned())
}

pub fn config_home() -> PathBuf {
    expand(env::var("XDG_CONFIG_HOME").unwrap_or_else(|_| "~/.config".into()))
}

pub fn data_home() -> PathBuf {
    expand(env::var("XDG_DATA_HOME").unwrap_or_else(|_| "~/.local/share".into()))
}

pub fn runtime_dir() -> Result<PathBuf> {
    env::var("XDG_RUNTIME_DIR")
        .map(PathBuf::from)
        .map_err(|_| LaeError::NoRuntimeDir)
}

pub fn lae_config_dir() -> PathBuf {
    config_home().join("lae")
}

pub fn lae_config_path() -> PathBuf {
    lae_config_dir().join("config.toml")
}

pub fn lae_data_dir() -> PathBuf {
    data_home().join("lae")
}

/// Resolve `[daemon].socket` from config to an absolute path.
///
/// Absolute paths and `~`-prefixed paths are expanded as-is. Bare filenames and
/// legacy values like `lae/daemon.sock` resolve under `~/.local/share/lae/`.
pub fn resolve_daemon_socket_path(configured: &str) -> PathBuf {
    let trimmed = configured.trim();
    if trimmed.starts_with('~') || trimmed.starts_with('/') {
        return expand(trimmed);
    }
    match trimmed {
        "daemon.sock" | "lae/daemon.sock" => lae_data_dir().join("daemon.sock"),
        other => lae_data_dir().join(other),
    }
}

pub fn lae_state_db() -> PathBuf {
    lae_data_dir().join("state.db")
}

pub fn lae_runtime_dir() -> Result<PathBuf> {
    Ok(runtime_dir()?.join("lae"))
}

/// User-local executables directory (`~/.local/bin`).
pub fn user_bin_dir() -> PathBuf {
    expand("~/.local/bin")
}

pub fn lae_waybar_file() -> Result<PathBuf> {
    Ok(lae_runtime_dir()?.join("waybar.json"))
}

pub fn lae_waybar_modules_cache() -> Result<PathBuf> {
    Ok(lae_runtime_dir()?.join("waybar-modules.json"))
}

pub fn ensure_parent(path: &Path) -> Result<()> {
    if let Some(parent) = path.parent() {
        std::fs::create_dir_all(parent).map_err(|source| LaeError::Write {
            path: parent.to_path_buf(),
            source,
        })?;
    }
    Ok(())
}
